### PR TITLE
feat(numa): bind numa if possible

### PR DESCRIPTION
**Pull request by li126com**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

To bind numa node

## Modification

try_bind_numa(launcher)

## Test
slurm:
image

torchrun:
image

check core:
image

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/initialize/launch.py [hunk 1]
```diff
@@ -474,3 +475,50 @@ def get_config_value(config, key, defalut):
     except KeyError:
         value = defalut
     return value
+
+
+def try_bind_numa(launcher):
+    try:
```
Comment (sunpengsdu) [inline internlm/initialize/launch.py:481]: 这里想想这么写会更好

```
try:
    import numa
    from numa import memory, schedule
    from pynvml.smi import nvidia_smi
except (AttributeError, ImportError) as e:
        logger.info(
            "Try bind numa failed! Package import error, if numa is not installed, "
            "please implement: pip install --upgrade py-libnuma"
        )

def try_bind_numa(launcher):

    # Early return if numa module not available
    if not numa:
        return
```

File: internlm/initialize/launch.py [hunk 2]
```diff
@@ -474,3 +475,50 @@ def get_config_value(config, key, defalut):
...
+        import numa
+        from numa import memory, schedule
+        from pynvml.smi import nvidia_smi
+    except (AttributeError, ImportError):
+        logger.info(
```
Comment (sunpengsdu) [inline internlm/initialize/launch.py:486]: 是不是只rank0 打印这个就行了

File: internlm/initialize/launch.py (hunk 1 above)
Comment (li126com) [reply internlm/initialize/launch.py:481]: 欸这样不会有点怪么，就是不启用try_bind_numa方法他也会去检查numa包

File: internlm/initialize/launch.py [hunk 3]
```diff
@@ -474,3 +475,60 @@ def get_config_value(config, key, defalut):
...
+            "Try bind numa failed! Package import error, if numa is not installed, "
+            "please implement: pip install --upgrade py-libnuma"
+        )
+else:
+    get_numa = True
+
+
+def try_bind_numa(launcher):
+    # Early return if numa module not available
+    if not get_numa:
```
Comment (SolenoidWGT) [inline internlm/initialize/launch.py:499]: Moving the code related to importing numa package into 'try_bind_numa' might make it easier to maintain.

File: train.py [hunk 4]
```diff
@@ -291,6 +291,9 @@ def main(args):
     args = parse_args()
     hostname = socket.gethostname()
 
+    if args.bind_numa:
```
Comment (SolenoidWGT) [inline train.py:294]: Try not to modify train.py, we should strive to make our modifications effective without requiring users to modify the interface code.

**Review by SolenoidWGT — APPROVED** [review]

**Pull request merged**